python/ray/air/util/data_batch_conversion.py

```diff
@@ -124,3 +94,63 @@ def convert_pandas_to_batch_type(data: pd.DataFrame, type: DataType) -> DataBatc
         raise ValueError(
             f"Received type {type}, but expected it to be one of {DataType}"
         )
+
+
+def _ndarray_to_column(arr: np.ndarray) -> Union[pd.Series, List[np.ndarray]]:
+    """Convert a NumPy ndarray into an appropriate column format for insertion into a
+    pandas DataFrame.
+
+    If conversion to a pandas Series fails (e.g. if the ndarray is multi-dimensional),
+    fall back to a list of NumPy ndarrays.
+    """
+    try:
+        # Try to convert to Series, falling back to a list conversion if this fails
+        # (e.g. if the ndarray is multi-dimensional.
+        return pd.Series(arr)
+    except ValueError:
+        return list(arr)
+
+
+def _cast_ndarray_columns_to_tensor_extension(df: pd.DataFrame) -> pd.DataFrame:
+    """Cast all NumPy ndarray columns in df to our tensor extension type, TensorArray."""
+    from ray.air.util.tensor_extensions.pandas import TensorArray
+    from ray.data.context import DatasetContext
+
+    ctx = DatasetContext.get_current()
```

Comment:
Can we move the context related logic to ray.data, and just have this flag be passed in as an arg?

Reply:
Hmm sure! Although Ray Train will still need to pass it in when calling within the predictors, is the Ray Train --> Ray Datasets dependency tenable?

Reply:
@amogkam AIR already has a dependency on Datasets, so why would we want to move this context check to the callers? It seems like that would amount to a lot of code duplication without much benefit.

Reply:
Hmm I thought AIR doesn't have a dependency on Datasets right now other than for type checking I believe?

These conversion functions are used for the Predictor -> Serve path, even when not using Datasets. Would it still make sense to have DatasetContext in that case?

Reply:
Sorry to clarify I mean just have an arg called `tensor_cast: bool`. In pandas_block, we can then inspect the Context and then specify this arg as necessary.

Reply:
>Hmm I thought AIR doesn't have a dependency on Datasets right now other than for type checking I believe?

Ah yeah, outside of the `air/util` modules, the dependency is only for type-checking, good point.

>These conversion functions are used for the Predictor -> Serve path, even when not using Datasets. Would it still make sense to have DatasetContext in that case?

The only method for disabling the automatic tensor extension casting is with the `DatasetContext`; do we need to also expose this as an AIR-level config?

Reply:
Ah sorry so there's 2 points:

1. I'm don't think it makes sense to have `DatasetContext` in these generic util functions that can be used outside of Datasets.
2. I also don't think we need this config for AIR in general, at least for now. For the Serve case, scale is small and we serialize/deserialize to json anyways, so the performance benefit from TensorArray should negligible?

Reply:
Agreed on the Serve case, but what about the batch prediction case? If there's no AIR config, then the user will need to use the Datasets context config to disable the tensor extension casting. I'm also not sure how we can support toggling the tensor extension casting for batch prediction when data batch conversion is done in the base predictor, which would also be used by Serve. 🤔

Reply:
For my own understanding how does a user set configs in DatasetContext currently? Is DatasetContext a global config that can be set?

Sorry my current understanding is that something like this would work, but let me know if this is incorrect
```
# Disable tensor conversion for all Dataset operations
ray.data.context.enable_tensor_conversion = False

data = ray.data.from_pandas(...)
predictor = BatchPredictor.from_checkpoint(...).predict(data)
```

Reply:
@amogkam Ping on this!

Reply:
Ah I see. 

So seems like there's two options?
1. Don't do anything in with regard to TensorArray in Predictor and rely on Datasets to convert it. The downside here is the extra copy
2. Use TensorArrays in Predictors. But we need a way to disable this. One possibility is to add a private arg to `Predictor.predict`, and then in `BatchPredictor` we can look at the `DatasetContext` to correctly set this arg.

Reply:
(or instead of a private arg, have it as a private attribute of Predictors)

Reply:
(2) sounds good to me since `BatchPredictor` is Datasets-native, and that would ensure consistent batch conversion across both training and prediction.

Will there be any issues of tensor extension casting happening for the preprocessors during training but not for the preprocessors during online inference using Serve? E.g. if they have a custom preprocessor that expects the upstream tensor extension columns (e.g. created via the imagery datasource) to be converted to NumPy ndarrays, that happens for the training preprocessors, but not for the Serve preprocessors, won't that be an issue? Or is this a non-issue because Serve has it's own batching logic that should already break this kind of dependence on the tensor extension representation?

Reply:
Also I'm thinking that it should be a private arg since `BatchPredictor` wraps other predictors, so having it as an overridable/settable private attribute of `Predictor` makes less sense to me.

Reply:
Ended up going with an instance variable that can be set by the `BatchPredictor` wrapper with a `predictor._set_cast_tensor_columns()` call.

Reply:
> Will there be any issues of tensor extension casting happening for the preprocessors during training but not for the preprocessors during online inference using Serve? E.g. if they have a custom preprocessor that expects the upstream tensor extension columns (e.g. created via the imagery datasource) to be converted to NumPy ndarrays, that happens for the training preprocessors, but not for the Serve preprocessors, won't that be an issue? Or is this a non-issue because Serve has it's own batching logic that should already break this kind of dependence on the tensor extension representation?

Hmm I would think the problem is the other way around? If users disable extension casting, and then write their preprocessors to work with TensorArrays, but for Serve TensorArrays are not used.

If extension casting is enabled, then users will be working with numpy arrays at all times right?

Reply:
>If users disable extension casting, and then write their preprocessors to work with TensorArrays, but for Serve, TensorArrays are not used.

That would also be an issue as well, I think.

>If extension casting is enabled, then users will be working with numpy arrays at all times right?

Correct, except for Serve since Serve doesn't do the automatic casting of tensor extension columns to ndarray columns, which is what I'm concerned about here. But I'm guessing that we're considering this a non-issue since it isn't currently possible to send online inference inputs to Serve over the wire that contain our tensor extension type?

Reply:
I don't think either of these cases are blockers, just making sure I understand any asymmetries here if users start seeing discrepancies between training preprocessing and online inference preprocessing.

Reply:
>  But I'm guessing that we're considering this a non-issue since it isn't currently possible to send online inference inputs to Serve over the wire that contain our tensor extension type?

Got it yep, this is my current understanding as well.